### DISCUSSION TITLE
ACK mechanism for cross-net messages in HC

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#119] [General]
Edit from @aakoshh's message in Slack: 
> The idea is to add a general acknowledgement mechanism for cross messages that always carry back to the source:
> - the outcome of the message (was it delivered, did it run out of gas midway) and
> - the unspent gas to where it was sent from. 
>
> This would have a few benefits:
> - the funds of the sender aren't spread across the system, wherever they sent the messages
> - they can always expect a positive/negative confirmation
> 
> I expect the latter could simplify the programming model of using cross messages. As I understand currently you only get feedback if the message failed to be delivered somewhere along the way, but not about successes, so it's difficult to say whether something is still in transit, or has it arrived yet and been successfully handled (without some explicit message sent back by the receiver, which again could fail). I suppose apps could monitor all chains along the way, but it would be more encapsulated if an ack/nack could be delivered back to the sender. Since the mechanism for failure signaling and the return of funds is already in the spec, we could just expand it to include the happy path as well, and (IMO) improve the UX of the handling of unspent gas.This could be a free service included in the price of delivering the message to its original destination in the first place. Since failure signaling is already free I assume, this would also fit into the picture.If I misunderstood the spec and this is already part of it, I'm sorry. (edited) 

See original motivation here: https://github.com/protocol/ConsensusLab/pull/113#discussion_r913091098

**Comment by adlrocha** [comment]
## The problem
As part of the discussion I think we should also revisit how the error message signalling works at a low-level when a cross-net message fails. For the MVP implementation, when a message fails, a cross-net message reverting the source and the destination of the original message is sent to: notify about the error, reverse potential changes in-transit, and return the message funds to its owner in the source subnet. The implementation didn't consider the cost of propagation of the error message back.

The only difference between cross-net messages and subnet messages is in the way in which they are executed. This means that the propagation of cross-net messages is impacted by the load and the fees paid in all of the subnets in the path to destination. Even more, each subnet may have its own gas model and related gas price mechanics. In the happy path, the source of the message pays for the fees to route a cross-net message, but who pays for the fees of the ACK cross-net message, and what if the ACK message fails or runs out of gas in-transit?

High-level any cross-net message can fail because: 
- Its execution fails in the destination subnet.
- It runs out-of-gas in-transit or it fails to be routed successfully.
- The message runs out of gas in-transit or in destination 
- (I may have left other error scenarios behind, let me know if this is the case).
And we should be able to notify of the failure successfully in all of these cases.

## Potential solutions
1. One approach would be to not send the ACK as a cross-net message but as a pubsub message. Each subnet could keep an event-log on-chain for the execution of cross-net messages in the SCA, and we could use the content resolution protocol to push/pull updates about cross-net messages between subnets. We could keep an independent pubsub topic for cross-net message event notification. The token leftovers in the failing message would be included in the originators address in the subnet where it failed and is up to them to recover the funds. The UX for users and applications for this approach is not great, and it worsens the problem of [crypto dust](https://www.gemini.com/cryptopedia/basics-of-crypto-dusting).
2. We could trigger an ACK cross-net message to the source to notify about the result of the execution. The problem here is who pays for the fees of that message to incentivize validators in-transit to include the message in their blocks. This would also increase the load in subnets as result of having to propagate these ACK messages:
  - We could make ACK messages free and mandatory, but how can we be sure that a validator in a subnet in-transit will actually include ACK messages.
  - We could have a fixed fee for the ACK that that the source needs to pay to send the message and that is never used (even if the message runs out of gas) and that is used to pay for the ACK message. The question here is, what is the right fee to ensure that the ACK makes it to the destination? Can we have a protocol level reward to validators to incentivize the routing of these ACK messages?

Ideas and feedback more than welcome. Let's try to flesh-out the best design for this :)

**Comment by vukolic** [comment]
not clear to me how this would exactly work. What if ACK message delivery fails (for the same/similar reasons as the original message), is there an ACK-ACK and NACK-ACK message, and so on ad infinitum?

**Reply by adlrocha** [reply]
That's the key. The MVP used that approach and the problem is that you could end up in an ACK/NACK loop of cross-net messages as you suggest. The work around to this that we came up with (and probably not the best solution, thus this discussion) is that failing ACKs are immediately discarded, there are no ACKs of ACKs to avoid loops.

**Reply by aakoshh** [reply]
I thought an ACK message has much less work to do than the original delivery, in that it doesn't need to resolve any potentially large message to include in the blocks, the packets should be small, the costs negligible. For this reason, it should not fail, and should not consume any gas. When it passes through a system other than the original source, it is just increasing the circulating supply of the subnet; only in the original subnet is it delivered to the sending actor, which can fail in some way, and can be charged some gas.

So, hopefully the looping can be avoided as it's never what I would want, as a user. 

As a user I'd like to be guaranteed that eventually I will get an ACK or a NACK. 

A NACK would come if 
* my message never reached the destination because it ran out of gas in transit
* my message reached the destination, where it failed
* my message may or may not have reached the destination, but the entire subnet where it went into failed or it was killed before it sent any feedback

**Reply by adlrocha** [reply]
> I thought an ACK message has much less work to do than the original delivery, in that it doesn't need to resolve any potentially large message to include in the blocks, the packets should be small, the costs negligible. For this reason, it should not fail, and should not consume any gas. When it passes through a system other than the original source, it is just increasing the circulating supply of the subnet; only in the original subnet is it delivered to the sending actor, which can fail in some way, and can be charged some gas.

If these ACK messages are free we may remove many of the potential sources of errors, however, they are still plain cross-net messages and they can fail. If they are routed as bottom-up messages, they have to be included in a checkpoint and resolved in the destination subnet (like any other cross-net message). What if the resolution fails because the subnet sending the ACK message stops being available? I am afraid there are still reasons why the message may fail, and we don't want to send and ACK-of-the-ACK. In these cases we should probably just discard the failing ACK message.

> A NACK would come if
>   - my message never reached the destination because it ran out of gas in transit
>   - my message reached the destination, where it failed
>   - my message may or may not have reached the destination, but the entire subnet where it went into failed or it was killed before it sent any feedback

I think this is a good start to start defining error codes for the ACK, this covers pretty much everything that can happen. I would probably add  one for "my message couldn't be executed because it failed to be resolved", If we agree we can probably start a small PR to add this to the spec.

**Reply by aakoshh** [reply]
> What if the resolution fails because the subnet sending the ACK message stops being available?

I thought that would be the 3rd NACK case. So you are saying that the original message has been successfully delivered to the destination, and the system is trying to send the ACK back, but that ACK cannot be resolved along the way, because the destination subnet is unavailable. 

From the perspective of the original sender, this cannot be distinguished from the message not being delivered at all. From the perspective of the lifecycle of the message, there should be exactly one ACK or NACK.

It is a fundamental question of how we can decide that a subnet has failed for good, and all messages that went to it must be NACK'd. It feels like there should be no coming back from that state, the subnet should be killed.

The IBC protocol messages has built-in timeouts, which might be relevant here as well, as there has to be some agreement over whether a message was or wasn't executed. Possibly the timeout is first enforced on the remote side, and then relayed back as a confirmation that this was the outcome.

> we don't want to send and ACK-of-the-ACK.

Definitely not! NACK'ing the ACK and ACK'ing the ACK back to the destination is not part of that lifecyle of the original message. It should be the concern of the HC machinery and treated accordingly.

> they are still plain cross-net messages

That's the crux, they probably shouldn't be. They should be part of the protocol, like ACK is part of TCP.

**Comment by aakoshh** [comment]
It's a good question on why the validators would propagate (N)ACKs for free. Maybe there could be some QoS level, governing the overall number of outstanding messages in the pipeline? Say, after 1000 messages passed on, the system is throttled until the feedback arrives, so if they want to get more fees for propagating cross messages, they must propagate feedback as well.

**Reply by aakoshh** [reply]
But it's probably easier to include a fixed fee for each level the ACK has to travel through, which can be claimed on the way back. 

I would assume that ACKs have to arrive in the same order as the cross messages, with some nonces tracking which one is supposed to come next from a given subnet. That way validators can't skip ACKs that are cheap, especially if they had the chance to inspect the message on the way out as well. 

Not sure if the fees should be locked on the way out, and consumed on the way back, or sent all the way there and back, never locked, trusting that the destination subnet will honestly send the value back including all remaining gas and unclaimed fees.

**Reply by adlrocha** [reply]
> Not sure if the fees should be locked on the way out, and consumed on the way back, or sent all the way there and back, never locked, trusting that the destination subnet will honestly send the value back including all remaining gas and unclaimed fees.

This may actually work, and we can use the CID of the original message to match ACK-message. So IIUC, your preference is to have ACKs as plain cross-net message (on-chain, maybe with a fixed fee), than a dedicated pubsub channel for ACKs at a transport level (off-chain, free but probably not reliable), right?

If we agree on this I think we can start drafting something on the spec and discuss over some baseline.

**Reply by jsoares** [reply]
Possibly dumb question, but what's the logic behind a fixed fee? Free seems to be a specific case of fixed fee, but the two retain the same problem: if there are more profitable messages to include, won't validators just do that (basically: starvation due to differing priorities/return)?

**Reply by aakoshh** [reply]
Good point. This is why I suggested the QoS approach where including an ACK is not a matter of choice, it's part of consensus, and without a steady stream of ACKs coming back, the subnet would not be eligible to receive more messages (and claim the fees) either, the source subnets would just stop propagating messages towards them. 

But it's just an intuition, I don't know how this should work exactly with an entire subtree. It feels to point in the direction of treating each `from` and `to` pair as a separate channel with their own throttling.

**Comment by aakoshh** [comment]
For possible inspiration: 
* https://github.com/cosmos/ibc/search?q=acknowledge 
* https://github.com/cosmos/ibc/tree/master/spec/app/ics-029-fee-payment

TL;DR is that in IBC (inspired by TCP/IP) there are channels opened (confirmed on both sides, can be ordered or unordered) between contracts on two different ledgers, but the relaying of chain-to-chain ledgers is left to permissionless and trustless relayers, who I think currently provide the service as a public good, for no fees from the message senders. The second issue points at discussions about possible incentive mechanisms.

**Comment by LefKok** [comment]
IIRC the last discussion was to persist the error where it occurs and let whoever cares (wallet provider probably) to do the propagation manually when reverting state is needed. 

The current discussion seems to claim that it is our job to reliable notify? I think this is philosophical but cluster-to-cluster communication is a very costly operation to force it on the system because someone missconfigured their cross-net transaction.

**Reply by aakoshh** [reply]
>  because someone missconfigured their cross-net transaction.

The spec says if the message runs out of gas in flight, the error is propagated back, and the remaining gas is returned. That sounds like the misconfigured cross-net transaction case you mention, yet backward propagation is promised.

I'm arguing for propagating success messages back to the source as well.

We could push this onto the wallet providers, but then it really becomes a whack-a-mole of failed ACK messages initiated by the wallets withering away midway and then nobody knows where to pick up the pieces. Also an ACK propagated automatically by the system would ensure that we have consensus over them, they could not be forged. 

>  cluster-to-cluster communication is a very costly operation to force it on the system

I thought this might be an opportunity to make the system more valuable for users by providing some guarantees, if possible. Like, you can use UDP which is completely unreliable, if it fits your use case, or you can use TCP which gives you many guarantees so that in your application layer you don't have to worry about those concerns or reivent the wheel on top of UDP (retries, out-of-order messages).
